plugins/core/src/terminal-service.ts

```diff
@@ -0,0 +1,73 @@
+import { ScryptedDeviceBase, StreamService } from "@scrypted/sdk";
+import type { IPty, spawn as ptySpawn } from 'node-pty-prebuilt-multiarch';
+import { createAsyncQueue } from '@scrypted/common/src/async-queue'
+
+export const TerminalServiceNativeId = 'terminalservice';
+
+export class TerminalService extends ScryptedDeviceBase implements StreamService {
+    async connectStream(input: AsyncGenerator<any, void>): Promise<AsyncGenerator<any, void>> {
+        const spawn = require('node-pty-prebuilt-multiarch').spawn as typeof ptySpawn;
+        const cp: IPty = spawn(process.env.SHELL, [], {});
+        const queue = createAsyncQueue<Buffer>();
+        cp.onExit(() => queue.end());
+
+        let bufferedLength = 0;
+        const MAX_BUFFERED_LENGTH = 64000;
+        cp.onData(async data => {
+            const buffer = Buffer.from(data);
+            bufferedLength += buffer.length;
+            const promise = queue.enqueue(buffer).then(() => bufferedLength -= buffer.length);
+            if (bufferedLength >= MAX_BUFFERED_LENGTH) {
+                cp.pause();
+                await promise;
+                if (bufferedLength < MAX_BUFFERED_LENGTH)
+                    cp.resume();
+            }
+        });
+
+        async function* generator() {
+            try {
+                while (true) {
+                    const buffers = queue.clear();
+                    if (buffers.length) {
+                        yield Buffer.concat(buffers);
+                        continue;
+                    }
+
+                    yield await queue.dequeue();
+                }
+            }
+            finally {
+                cp.kill();
+            }
+        }
+
+        setTimeout(async () => {
```

Comment:
I think I had this setTimeout pattern before due to some weird issue with the event loop on engine.io. I think this can be replaced with an in place

(async() => {
 ... code
})()